### DISCUSSION TITLE
[back] 소켓을 몇 개 만들 지에 대한 논의

### DISCUSSION THREAD

**Discussion by JuneParkCode** [3DPong/transcendence#22] [General]
저희 서비스에서는 소켓을 사용하는 서비스가 총 3개가 있습니다. [ 채팅, 게임, 알람 ]. 3가지 서비스에서 같은 소켓을 사용할 수도, 각자 다른 소켓을 사용할 수도 있습니다. 그 각자의 장단점이 있기 때문에, 이유를 명확하게 하고 분리를 하던, 통합을 하면 좋겠습니다.
@kyu-baek @csstring 

아래는 chat gpt에게 문의한 결과입니다.
---

You have two main options when using Socket.IO with NestJS for your web game:

### Use a single socket for all services:
In this approach, you would use a single Socket.IO instance to handle all communication between the client and server for alarms, game networking, and chatting. This can simplify your implementation, as you only have to manage one connection per client.

**Pros:**
- Simpler implementation
- Less overhead (fewer connections to manage)
- Easier to maintain and reason about

**Cons:**
- Less modular and flexible
- Potential performance issues if one service requires a lot of bandwidth or has high latency requirements
- Can be harder to scale if each service has different scaling needs

### Use separate sockets for each service:
In this approach, you would create separate Socket.IO instances for each service (alarms, game networking, and chatting). This allows you to isolate the services and optimize them independently.

**Pros:**
- More modular and flexible
- Easier to scale each service independently
- Better isolation of concerns, making it easier to troubleshoot and optimize individual services

**Cons:**
- More complex implementation
- Increased overhead (more connections to manage)
- Potentially more challenging to maintain

### Summary
In summary, if your services have distinct requirements and need to scale independently, using separate sockets for each service can be a good choice. However, if your services are closely related and have similar requirements, using a single socket might be simpler and easier to manage.

Ultimately, the best choice depends on your specific use case and requirements. It's important to consider the trade-offs and make a decision based on your project's needs.

**Comment by JuneParkCode** [comment]
관점을 어떻게 잡냐에 따라 조금 달라질 것 같습니다. 

만약 상용 서비스가 되고, 유저가 많이 늘어날 때를 고려하면 각자를 분리하여 구성하는 것이 좋은 방법이라고 생각합니다. (물론 과제의 목적에서는 의미없지만)
이유는 확장성에 있습니다. 사용자가 많아지면, 서버의 분리가 필요해질 것 같습니다. 아마, api서버, 게임 서버, 채팅 서버 정도로 분리될 것 같습니다. 이 때, 동일 소켓에서 모든 정보를 주고받는 방식은 오히려 독이 됩니다. 반면 각 서비스마다 소켓을 따로 팠다면 해당 서버에서 처리를 하게 되니 편리해질 것이구요.  이 경우에서는 아마, alarm socket 을 api 서버 쪽에 물려두고, 해당 소켓의 상태로 online, offline 상태를 점검하면 될 것 같네요.

비록 과제이긴 하지만, session을 redis에 올리는 이유도 비슷한 이유에서 하는 만큼, 상용 서비스가 됐을 때를 고려하고 싶어서 저는 이렇게 생각해봤습니다. 다른 분들은 어떻게 생각하시나요?

**Reply by kyu-baek** [reply]
기능에 따라 소켓을 분리하면 기능적인 이점도 있고, 실무적으로도 각자 개발하고 테스트 할 때 편하기 때문에 소켓을 세 개 나누어 사용하는 것에 동의 합니다. 아니면 ( 채팅 + 알림 && 게임 ) or (채팅 && 게임 + 알림) 식으로 병합해도 괜찮다고 생각합니다. 
하지만 채팅과 게임은 room 형식으로 반복해서 생겨났다 사라지기 때문에 섞이는 것 보단 분리하는 게 좋다고 생각합니다. 

알림 기능의 확장성을 생각해서 따로 분리하시거나, 합치거나 결정해주시면 감사하겠습니다.

**Reply by csstring** [reply]
기능 별로 module을 나누고 controller를 생성하는 것과 같은 이유로 module단위로 gateway를 생성하는게 맞는것 같습니다. 특히 socket부분은 개발 과정에 있어서 수시로 코드가 바뀌는 부분이라 개발,유지보수 할때 이점을 가져갈 수 있을거 같습니다
